### PR TITLE
Allow first character of OCAMLPARAM to specify an alternative separator

### PR DESCRIPTION
**Pull request by lefessan**
In the current implementation of `OCAMLPARAM`, there is no way to escape a comma in the arguments (for example, if you want to pass an option to the linker in ccopts). This PR provides such a way, by allowing the user to specify an alternative separator, instead of comma, as first character of `OCAMLPARAM`. The alternative separator must be in ":|; ".

For example, `OCAMLPARAM='x=1,_,y=1'` can also be written `OCAMLPARAM='|x=1|_|y=1'`.

A small bug fix also is included, by adding a `ccopt` option, in addition to `ccopts`.

### CODE REVIEW

**Comment by ghost** [comment]
Personally, I find `OCAMLPARAM` already painful to use. If you are in a shell script and want to extend it you have to first check whether it's empty or not. Now you'll also have to check what syntax it's using. 

Why not simply use one environment variable per setting, like what opam does?

**Comment by lefessan** [comment]
@diml I would be curious to know when you need to do that. The intention for `OCAMLPARAM` was that it should only be used to manually modify a build, so cascading effects should not appear.

Anyway, this PR is not about replacing `OCAMLPARAM` by another mechanism, but to fix the current limitation that prevents commas in arguments. It should be done even if we decide to think about something else on the long term.

**Comment by ghost** [comment]
> @diml I would be curious to know when you need to do that. The intention for OCAMLPARAM was that it should only be used to manually modify a build, so cascading effects should not appear.

I do that in jbuilder, to force the compiler to print with colors even though the output is redirected. jbuilder then strips colors itself depending on its own output.

**Comment by dbuenzli** [comment]
@diml there's now https://github.com/ocaml/ocaml/pull/1098 and/or you could simply use `-color always`

**Comment by ghost** [comment]
Indeed, this should simplify things

**Comment by lefessan** [comment]
If there is no opposition, I will commit this feature by the end of the week.

**Comment by shindere** [comment]
Fabrice Le Fessant (2017/05/15 04:38 -0700):
> If there is no opposition, I will commit this feature by the end of
> the week.

At the moment this PR does not update the documentation, but perhaps you
intended to do so once you were sure there is a consensus about the
feature?

On this matter: it seems to me that there is a very important gap
between what is documented in manual/manual/cmds/runtime.etex and what
is really supported in driver/compenv.ml, functions parse_args and
read_one_param. Is it on purpose that so many possible parameters are
not documented, or perhaps they are documented in another place?

Actually, the presentation of OCAMLRUNPARAM is really misleading because
it insists that parameter names are _letters_, which is not the case for
many of them (I guess this is so for historical reasons, but it would
probably diserve a good refresh).

I realise this PR is not responsible for all of this, I just wanted to
mention the fact and perhaps avoid that this PR makes the gap between
the documentation and the real behaviour even bigger it already is or
seems to be, if I missed something.

@Octachron: perhaps you know something regarding the documentation?

Coming to the proposed feature itself, I have to say I do not fully
understand why it would be useful. For the given example, couldn't the
same result be reached by setting LDFLAGS or something similar?


Even if that cannot be done, I also like the idea of using separate
environemnt variables, especially for new features.

One could even imagine to extend the framework that handles command-line
options so that for each defined option it becomes possible to specify
an environment variable that corresponds to this option and overloads
it, if defined.

**Comment by lefessan** [comment]
Thanks Sebastien for your comment. I updated the manual, the description of `OCAMLPARAM` was pretty vague, so I made it a bit more explicit, still without specifying the names of the arguments.

**Comment by shindere** [comment]
Fabrice Le Fessant (2017/05/16 06:56 -0700):
> Thanks Sebastien for your comment. I updated the manual, the
> description of `OCAMLPARAM` was pretty vague, so I made it a bit more
> explicit, still without specifying the names of the arguments.

Thanks, Fabrice, this indeed makes the documentation more
understandable.

Regarding your text, I'd say "interpreted" or "taken into account"
rather than inserted / executed. Perhaps our english native speaker
friends can comment on this.

"should be executed before parsing the arguments": I'd say the
_command-line_ arguments, here, to make things yet clearer.

Have you seen the sentence that starts with "A parameter specification
is an option letter followed by an "=" sign,"? That is confising, I
think, because names can actually be something else than just a letter.

Also: doyou intend to address my other comments?

To be honest I find the documentation update nice but am still skeptical
about the feature you suggest so I find it a bit a pity to have them in
the same PR.

**Comment by Octachron** [comment]
@shindere Concerning the documentation of `OCAMLPARAM`, the documentation itself is quite recent cf. [mantis:7007](https://caml.inria.fr/mantis/view.php?id=7007). Another point is that `OCAMLRUNPARAM` is indeed limited to one-letter parameters (see [byterun/startup_aux](https://github.com/ocaml/ocaml/blob/e0749dec9701b21c64c1c607120918c550e50053/byterun/startup_aux.c#L80)), contrarily to `OCAMLPARAM`.

**Comment by rneswold** [comment]
@shindere :
> Regarding your text, I'd say "interpreted" or "taken into account" rather than inserted / executed. Perhaps our english native speaker friends can comment on this.

I agree with your recommendation: I'd replace "inserted" with "interpreted". For "executed", I'd lean more towards "processed". Or maybe the sentence fragment:

`means that "a=x" should be executed before parsing the arguments, and "b=y" after.`

could be rewritten as

`means that "a=x" should be parsed before the command line, and "b=y" after. `

**Comment by shindere** [comment]
Thanks, @octachron !
I think I mixed up OCAMLPARAM and OCAMLRUNPARAM, sorry

**Comment by SkySkimmer** [comment]
Why not provide a way to escape commas instead?

**Comment by lefessan** [comment]
@SkySkimmer For two reasons:
* It is quite standard in the Unix world to use the first char as a separator;
* Escaping is a complex mechanism, especially if you are operating directly from the shell. Also it creates other questions: how do you escape ? Double comma or backslash comma, in which case you also need to escape backslash...

**Comment by shindere** [comment]
Fabrice Le Fessant (2017/05/22 00:17 -0700):
> * It is quite standard in the Unix world to use the first char as a
> separator;

Ah? Which examples do you have in mind, actually?

**Comment by lefessan** [comment]
> Fabrice Le Fessant (2017/05/22 00:17 -0700):
>  * It is quite standard in the Unix world to use the first char as a
>  separator;
> Ah? Which examples do you have in mind, actually?

For example, `sed` uses any character following the `s` command as the separator. In LaTeX, `\verb` uses also the next character as separator. Although I don't have other examples in mind, I think it is a common solution to avoid escaping.

**Comment by alainfrisch** [comment]
Why do you restrict to a fixed set of separator?  Any character which is not a letter and not '_' would work, no?

**Comment by lefessan** [comment]
For backward compatibility: we want the separator to be something that was not used before. Indeed, the set would be bigger than now, probably anything except letters, '_' and '='.

File: driver/compenv.ml [hunk 1]
```diff
@@ -106,14 +106,24 @@ type readenv_position =
...
 
 let parse_args s =
-  let args = String.split_on_char ',' s in
+  let args =
+    let len = String.length s in
+    if len = 0 then []
+    else
+      (* allow first char to specify an alternative separator in ":|; " *)
+      match s.[0] with
+      | ( ':' | '|' | ';' | ' ' ) as c ->
```
Comment (damiendoligez) [inline driver/compenv.ml]: You should also allow `,` here.

File: driver/compenv.ml [hunk 2]
```diff
@@ -106,14 +106,24 @@ type readenv_position =
...
+  in
   let rec iter is_after args before after =
     match args with
       [] ->
       if not is_after then
         raise (SyntaxError "no '_' separator found")
       else
-      (List.rev before, List.rev after)
+        (List.rev before, List.rev after)
+    | "" :: tail -> iter is_after tail before after
```
Comment (damiendoligez) [inline driver/compenv.ml:127]: Note: you are now allowing empty elements in the list.

File: driver/compenv.ml [hunk 3]
```diff
@@ -358,7 +368,9 @@ let read_one_param ppf position name v =
         ccobjs := Misc.rev_split_words v @ !ccobjs
     end
 
-  | "ccopts" ->
+  | "ccopt"
+  | "ccopts"
+    ->
```
Comment (damiendoligez) [inline driver/compenv.ml:374]: Looks like an unrelated change that should be in another PR.

File: manual/manual/cmds/native.etex [hunk 4]
```diff
@@ -135,8 +135,13 @@ development work, not in released packages.
...
 \begin{options}
 \item["OCAMLPARAM" \rm(environment variable)]
-Arguments that will be inserted before or after the arguments from the
-command line.
+A set of arguments that will be inserted before or after the arguments from
+the command line. Arguments are specified in a comma-separated list
+of "name=value" pairs. A "_" is used to specify the position of
+the command line arguments, i.e. "a=x,_,b=y" means that "a=x" should be
+executed before parsing the arguments, and "b=y" after. Finally,
+an alternative separator than the comma can be specified as the
```
Comment (damiendoligez) [inline manual/manual/cmds/native.etex]: I don't think this is idiomatic English. Just drop "than the comma".

File: manual/manual/cmds/native.etex [hunk 5]
```diff
@@ -135,8 +135,13 @@ development work, not in released packages.
...
+first character of the string, within the set ":|; ".
```
Comment (damiendoligez) [inline manual/manual/cmds/native.etex]: Add `,` here too.

File: manual/manual/cmds/comp.etex [hunk 6]
```diff
@@ -122,8 +122,13 @@ development work, not in released packages.
...
 \begin{options}
 \item["OCAMLPARAM" \rm(environment variable)]
-Arguments that will be inserted before or after the arguments from the
-command line.
+A set of arguments that will be inserted before or after the arguments from
+the command line. Arguments are specified in a comma-separated list
+of "name=value" pairs. A "_" is used to specify the position of
+the command line arguments, i.e. "a=x,_,b=y" means that "a=x" should be
+executed before parsing the arguments, and "b=y" after. Finally,
+an alternative separator than the comma can be specified as the
```
Comment (damiendoligez) [inline manual/manual/cmds/comp.etex]: See below.

**Review by damiendoligez — CHANGES_REQUESTED** [review]
You should add a Changes entry.

**Comment by mshinwell** [comment]
@lefessan Could you deal with @damiendoligez 's comments so we can finish this off?

**Comment by xavierleroy** [comment]
Requested changes have not been performed.  Plus, I'd rather have proper escapes rather than this `\verb`-like trick.  Shall we finish this PR or close it?

File: driver/compenv.ml (hunk 2 above)
Comment (lefessan) [reply driver/compenv.ml:127]: Yes, it was actually my intent.

File: driver/compenv.ml (hunk 3 above)
Comment (lefessan) [reply driver/compenv.ml:374]: Is it really necessary to open a new PR ? This change is only to be consistent with the names of arguments.

**Comment by lefessan** [comment]
Requested changes performed.

**Comment by lefessan** [comment]
Looks like Appveyor is not working, again

File: Changes [hunk 7]
```diff
@@ -14,6 +14,10 @@ be mentioned in the 4.06 section below instead of here.)
 
 ### Compiler user-interface and warnings:
 
+- GPR#1166: In OCAMLPARAM, an alternative separator can be specified as
+  first character (instead of comma) in the set ":|; "
```
Comment (damiendoligez) [inline Changes]: Add the comma here too, then we're good to merge.

**Review by damiendoligez — APPROVED** [review]
Good to merge after one small fix.

**Comment by lefessan** [comment]
Small fix done, rebased in one commit.

**Comment by damiendoligez** [comment]
Merged into trunk. Do **not** cherry-pick to 4.06.
